### PR TITLE
Allow setting custom BodyExtractor to the AbstractWebClientReactiveOAuth2AccessTokenResponseClient

### PR DESCRIPTION
**Pull request by bishoybasily**
A setter added to set the `BodyExtractor` inside the `AbstractWebClientReactiveOAuth2AccessTokenResponseClient`
Closes gh-10260

### CODE REVIEW

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractWebClientReactiveOAuth2AccessTokenResponseClient.java [hunk 1]
```diff
@@ -291,4 +294,14 @@ public final void addHeadersConverter(Converter<T, HttpHeaders> headersConverter
...
 	}
 
+	/**
+	 * Sets the {@link BodyExtractor} that will be used to decode the {@link OAuth2AccessTokenResponse}
+	 *
+	 * @param bodyExtractor the {@link BodyExtractor} that will be used to decode the {@link OAuth2AccessTokenResponse}
+	 * @since 5.6
+	 */
+	public void setBodyExtractor(BodyExtractor<Mono<OAuth2AccessTokenResponse>, ReactiveHttpInputMessage> bodyExtractor) {
+		this.bodyExtractor = bodyExtractor;
```
Comment (jzheaux) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/endpoint/AbstractWebClientReactiveOAuth2AccessTokenResponseClient.java:304]: Will you please add a null check here? `Assert.notNull` is handy.

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks, @bishoybasily!

Please see my inline comments. In addition, will you please add tests that confirm this method works? I think there should be one in each `AbstractWebClientReactiveOAuth2AccessTokenResponseClient` implementation's test classes.

Once you are ready, will you please squash your commits and ensure that the commit message ends with "Closes gh-10260" like so:

```
Add setBodyExtractor

Closes gh-10260
```

**Comment by bishoybasily** [comment]
Hi @jzheaux,
all done, please let me know if something else is missing

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for the quick turnaround, @bishoybasily!

In addition to my inline comments, will you please double-check that you have a successful run of

```bash
./gradlew format check
```

and make the needed changes? And finally, remember to squash your commits when you are ready.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/endpoint/WebClientReactiveAuthorizationCodeTokenResponseClientTests.java [hunk 2]
```diff
@@ -409,4 +416,82 @@ public void convertWhenHeadersConverterSetThenCalled() throws Exception {
...
+				+ "   \"expires_in\": \"3600\",\n"
+				+ "   \"scope\": \"openid profile\",\n"
+				+ "   \"refresh_token\": \"refresh-token-1234\",\n"
+				+ "   \"custom_parameter_1\": \"custom-value-1\",\n"
+				+ "   \"custom_parameter_2\": \"custom-value-2\"\n"
+				+ "}\n";
+		// @formatter:on
+
+		WebClientReactiveAuthorizationCodeTokenResponseClient customClient = new WebClientReactiveAuthorizationCodeTokenResponseClient();
+		customClient.setBodyExtractor((inputMessage, context) -> {
```
[the remark this answers is not in the document]
Comment (bishoybasily) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/endpoint/WebClientReactiveAuthorizationCodeTokenResponseClientTests.java]: replaced with a mock 👍

**Comment by bishoybasily** [comment]
Hi @jzheaux 

Done!

- 1. - Impl. replaced with mock in tests
- 2. - CheckStyle errors fixed

**Review by jzheaux — APPROVED** [review]

**Comment by jzheaux** [comment]
Thanks, @bishoybasily, this is now merged into `main` with 860690491aad0ee16ea3b49ab267995b4a30d90f

**Comment by bishoybasily** [comment]
Thanks, @jzheaux

**Pull request closed without merging**